<a href="https://colab.research.google.com/github/timothy-shiwprasad/COMP6905_assigment_1/blob/main/Assigment_1_Question3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Question 3

## Key Information

x<sub>S</sub> = Acres for Starter Cottage Development

x<sub>F</sub> = Acres for Family Home Development

x<sub>T</sub> = Acres for Trainer Townhouse Development

x<sub>A</sub> = Acres for Apartment Development

### Construction Cost

CC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 900x<sub>S</sub> + 1250x<sub>F</sub> + 1500x<sub>T</sub> + x<sub>A</sub>

### Peak Electricity Requirement

PER(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 22x<sub>S</sub> + 30x<sub>F</sub> + 45x<sub>T</sub> + 70x<sub>A</sub>

### Water Requirement

WR(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 2.0x<sub>S</sub> + 2.8x<sub>F</sub> + 3.6x<sub>T</sub> + 5.0x<sub>A</sub>

### Utility Connection Assessment per Acre

UCA(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 120x<sub>S</sub> + 160x<sub>F</sub> + 190x<sub>T</sub> + 260x<sub>A</sub>

### Utility Connection

UC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = max{3000, UCA(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>)}

### Expected Net Return

ENR(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 500x<sub>S</sub> + 640x<sub>F</sub> + 790x<sub>T</sub> + 180x<sub>A</sub> - UC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>)

### Subject To

* Acres < 18
* Electricity ≤ 750 kW
* Water per day ≤ 60000
* x<sub>S</sub> ≥ 0.4(18)
* Construction Cost ≤ 26000


# Question 3 A

## Decision Variables
x<sub>S</sub> = Acres allocated to Starter Cottage Development

x<sub>F</sub> = Acres allocated to Family Home Development

x<sub>T</sub> = Acres allocated to Trainer Townhouse Development

x<sub>A</sub> = Acres for Apartment Development


## Objective Function
This objective functions is created by the combination of three componets mentioned in the problem.  
1. Expected Net Return: ENR(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 500x<sub>S</sub> + 640x<sub>F</sub> + 790x<sub>T</sub> + 180x<sub>A</sub>

2. Utility Connection assesment per acre: UCA(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 120x<sub>S</sub> + 160x<sub>F</sub> + 190x<sub>T</sub> + 260x<sub>A</sub>

3. Actual Utility Conection cost : AUC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = max{3000, UCA(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>)}

Final Objective Function:

Actual Net Return : ENR(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) = 500x<sub>S</sub> + 640x<sub>F</sub> + 790x<sub>T</sub> + 180x<sub>A</sub> - AUC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>)

For Simplicity:
Let x = (x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>)

ANR(x) - AUC(x)

Objective: max ANR(x) - AUC(x) ,x >= 0


## Constrains
- x<sub>S</sub> + x<sub>F</sub> + x<sub>T</sub> + x<sub>A</sub> <= 18
- 22x<sub>S</sub> + 30x<sub>F</sub> + 45x<sub>T</sub> + 70x<sub>A</sub> <= 750
- 2.0x<sub>S</sub> + 2.8x<sub>F</sub> + 3.6x<sub>T</sub> + 5.0x<sub>A</sub> <= 60
- 0.6x<sub>S</sub> - 0.4<sub>F</sub> - 0.4x<sub>T</sub> - 0.4x<sub>A</sub> >= 0
- 900x<sub>S</sub> + 1250x<sub>F</sub> + 1500x<sub>T</sub> + x<sub>A</sub> - AUC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) <= 26000







# Question B

We will replace the max function AUC(x) and define it as a constrain.

u = AUC(x)

## New Constrains
- u >= 3000
- u >= 120x<sub>S</sub> + 160x<sub>F</sub> + 190x<sub>T</sub> + 260x<sub>A</sub>

New Objective Function:
Z(x)  = 500x<sub>S</sub> + 640x<sub>F</sub> + 790x<sub>T</sub> + 180x<sub>A</sub> - u


## Constrains
- x<sub>S</sub> + x<sub>F</sub> + x<sub>T</sub> + x<sub>A</sub> <= 18
- 22x<sub>S</sub> + 30x<sub>F</sub> + 45x<sub>T</sub> + 70x<sub>A</sub> <= 750
- 2.0x<sub>S</sub> + 2.8x<sub>F</sub> + 3.6x<sub>T</sub> + 5.0x<sub>A</sub> <= 60
- 0.6x<sub>S</sub> - 0.4<sub>F</sub> - 0.4x<sub>T</sub> - 0.4x<sub>A</sub> >= 0
- 900x<sub>S</sub> + 1250x<sub>F</sub> + 1500x<sub>T</sub> + x<sub>A</sub> - AUC(x<sub>S</sub>, x<sub>F</sub>, x<sub>T</sub>, x<sub>A</sub>) <= 26000
- u >= 3000
- u >= 120x<sub>S</sub> + 160x<sub>F</sub> + 190x<sub>T</sub> + 260x<sub>A</sub>

### Reasoning
The constrain forces u to be both  at least 3000 and AUC(x). Since u is being subtracted, trhe model will chose a U value that is smallest and matches the 2 constrains. This in return would ensure that we are maximizing the Net return.



In [12]:
using Pkg; Pkg.add("JuMP"); Pkg.add("Clp")

using JuMP
using Clp

model = Model(Clp.Optimizer)

# Decision variables In acres
@variable(model, xS >= 0)
@variable(model, xF >= 0)
@variable(model, xT >= 0)
@variable(model, xA >= 0)

# Auxiliary variable for utility charge in thousands
@variable(model, u >= 0)

# Constraints
@constraint(model, land, xS + xF + xT + xA <= 18)
@constraint(model, elec, 22xS + 30xF + 45xT + 70xA <= 750)
@constraint(model, water, 2.0xS + 2.8xF + 3.6xT + 5.0xA <= 60)
@constraint(model, starter_share, 0.6xS - 0.4xF - 0.4xT - 0.4xA >= 0)
@constraint(model, budget, 900xS + 1250xF + 1500xT + 2200xA + u <= 26000)
@constraint(model, u_min, u >= 3000)
@constraint(model, u_assess, u >= 120xS + 160xF + 190xT + 260xA)

# Objective: maximize net return after paying utility charge
@objective(model, Max, 500xS + 640xF + 790xT + 1080xA - u)

optimize!(model)

   Resolving package versions...
     Project No packages added to or removed from `~/.julia/environments/v1.12/Project.toml`
    Manifest No packages added to or removed from `~/.julia/environments/v1.12/Manifest.toml`
   Resolving package versions...
     Project No packages added to or removed from `~/.julia/environments/v1.12/Project.toml`
    Manifest No packages added to or removed from `~/.julia/environments/v1.12/Manifest.toml`


Coin0506I Presolve 6 (-1) rows, 5 (0) columns and 26 (-1) elements
Clp0006I 0  Obj -3000 Dual inf 3010 (4)
Clp0006I 3  Obj 9264.5714
Clp0000I Optimal - objective value 9264.5714
Coin0511I After Postsolve, objective 9264.5714, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 9264.571429 - 3 iterations time 0.002, Presolve 0.00


0.41428571428571415

In [14]:
# Optimal Acres for each development chosen to maximize Expected Net Return
xS_opt = value(xS)
xF_opt = value(xF)
xT_opt = value(xT)
xA_opt = value(xA)
u_opt  = value(u)
obj_opt = objective_value(model)

println("The Optimal Acres for each development type is as follows: Starter Cottage Development: $(xS_opt) , Family House Development : $(xF_opt) , Trainer Townhouse Development: $(xT_opt), Apartment Development : $(xA_opt)")

println("The maximum Expected Net Return (In Thousands): $(obj_opt)")

The Oprimal Acres for each development type is as follows: Starter Cottage Development: 7.200000000000001 , Family House Development : 0.0 , Trainer Townhouse Development: 10.342857142857138, Apartment Development : 0.45714285714286085
The maximum Expected Net Return (In Thousands): 9264.57142857143


In [20]:
land_lhs = value(xS + xF + xT + xA)
elec_lhs = value(22xS + 30xF + 45xT + 70xA)
water_lhs = value(2.0xS + 2.8xF + 3.6xT + 5.0xA)
budget_lhs = value(900xS + 1250xF + 1500xT + 2200xA + u)

println("Contraints bindings at the Optimal")
println("Acres Constraints: $(land_lhs)")
println("Electricity Contraints: $(elec_lhs)")
println("Water Constraints: $(water_lhs)")
println("Budget Constrains: $budget_lhs")

Contraints bindings at the Optimal
Acres Constraints: 18.0
Electricity Contraints: 655.8285714285714
Water Constraints: 53.92
Budget Constrains: 26000.0


# Question C Findings

## Values
x<sub>S</sub> = 7.2

x<sub>F</sub> = 0.0

x<sub>T</sub> = 10.342857142857138

x<sub>A</sub> = 0.45714285714286085

u = 3000

Maximum Expected Net Income ($000) = $9364.57142857143

The constraints for Acres and Budget are binding at the optimum. The values
are 18 (acres) and $26000 ($000) respectively which are the maximum values they can be.

## Interpretations

In order to take advantage of the substantial increase in population and visitors we will have to allocate our resources to the following residential developemnts. The investement in these areas will ensure to maximize our Expected Net Returns while staying within our land, electricity, water and budget constraints.Our allocation also prioritizing Starter Cottage Developement with at least 40% of the development being Starter Cottage Development. We should prioritise the development of Starter Cottage Developement and Trainer Townhouse Development, allocating 7.2 acres and 10.3 acres respectively. We then allocate 0.46 acres to Apartment Developments. Unfortunately in order to maximise our returns Family Homes will not be built. Overall the majority of the developments in Starter Cottages and Trainer Townhouses will be the most beneficial as our recent surge in population are a result of Red's victory and many trainers are coming to either duplicate his journey , train here or in hope of challenging him.



In [27]:
# Question D
land_dual = shadow_price(land)
println(land_dual)
elec_dual = shadow_price(elec)
println(elec_dual)
water_dual = shadow_price(water)
println(water_dual)
budget_dual = shadow_price(budget)
println(budget_dual)

152.00000000000023
-0.0
-0.0
0.41428571428571415


## Question D
# Interpretation of Shadow Prices
The Shadow Prices for the contraints are as followed:

- Land : 152.00
- Electricty: 0.0
- Water: 0.0
- Budget: 0.414

The shadow prices for electricity and water are 0.0 , since there is still additional resources avaliable for electricity and water (slack) an increase in the KW of electricy being genereated or a increase of gallons of water provided will not increase our objective of maximizing our Net Returns. We essentially have more than enough.

On the other hand the shadow prices for Land and Budget are 152 and 0.414 respectively. These two contraints are binding and based on their shadow prices we can expect an increase of 152 `($000)` to our net return for every increase unit increase of acre. The same for the budget. A single  increase of  1 `($000)`  to the budget will result in an 0.414 `($000)` increase to our Net Return.

In [30]:
# Question E

# Calculation the increase of the Net Return using the shadown price of the Budget

increase_to_net_return = budget_dual * 500
println("Predicted Increase: $(increase_to_net_return)")
objective_with_budget_increase = obj_opt + increase_to_net_return

println("Budget Increase Objective: $(objective_with_budget_increase)")


Predicted Increase: 207.14285714285708
Budget Increase Objective: 9471.714285714286


In [36]:
## Adjustment to the Budget constraint before running the Model.

# delete and re-add in a single block
delete(model, budget)
budget = @constraint(model, 900xS + 1250xF + 1500xT + 2200xA + u <= 26500)
optimize!(model)

obj_opt = objective_value(model)

println("New Objective: $(obj_opt)")

New Objective: 9471.714285714286
Coin0506I Presolve 6 (-1) rows, 5 (0) columns and 26 (-1) elements
Clp0006I 0  Obj -3000 Dual inf 3010 (4)
Clp0006I 3  Obj 9471.7143
Clp0000I Optimal - objective value 9471.7143
Coin0511I After Postsolve, objective 9471.7143, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 9471.714286 - 3 iterations time 0.002, Presolve 0.00


## Querstion E continued
## Comparing the Objective generated using the increase budget and Shadow Price.

Using Shadow Price : 9471.714285714286 ($000)

Using Model with increased Budget: 9471.714285714286 ($000)


The Values are the same.